### PR TITLE
Nuke huge ResourceProperties sum type

### PR DESCRIPTION
**Pull request by jdreaver**
This PR nukes the huge `ResourceProperties` sum type. Instead of that, I created a `ResourceProperties` record holding the Type and Properties for each resource, along with a `ToResourceProperties` class for syntactic convenience. Instead of making a resource, finding the right constructor in `ResourceProperties` to use, and then using that in `resource "MyResource" _`, you can just stuff your resource into `resource "MyResource" _`.

The reason I did this is to reduce compilation time/memory, and to reduce the resulting binary size of libraries using `stratosphere` (see #95). Also, since we decided to nuke `FromJSON` (see #118), we really don't need this big sum type.

Luckily the changes to the examples are early on in this diff so you can see the difference. For example, this is all that has to change for a resource from the RDS instance example:

```diff
@@ -32,7 +32,6 @@ dbTemplate =
 rdsMaster :: Resource
 rdsMaster =
   resource "RDSMaster" $
-  RDSDBInstanceProperties $
   rdsdbInstance
   "db.t2.micro"
```

`resource` now has type `resource :: (ToResourceProperties a) => Text -> a -> Resource`, and each resource is an instance of `ToResourceProperties`.

This is definitely an annoying breaking change, but hopefully it is an extremely easy one to make for our users.

cc any users of `stratosphere` I can think of in case anyone has any comments:
- @cdparks 
- @MichaelXavier 
- @ababkin 
- @timmytofu 
- @andreyk0
- @utdemir
- @Znack
- @amar47shah

### CODE REVIEW

**Comment by andreyk0** [comment]
This looks great, thanks!
BTW .. since you are making big changes ... have you ever considered wrapping all of those "text" parameters into newtypes? 

We wrapped text into things like 
```

File: library/Stratosphere/ResourceProperties.hs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  ) where
+
+import Data.Aeson
+import Data.Aeson.Types (Pair)
+import Data.Text (Text)
+
+data ResourceProperties
+  = ResourceProperties
+  { resourcePropertiesType :: Text
+  , resourcePropertiesProperties :: Value -- Maybe use Object here?
```
Comment (cdparks) [inline library/Stratosphere/ResourceProperties.hs]: `Value` is fine, best-case scenario with `Object` is that you have to explode if something doesn't return a `HashMap`. Also remember `-XStrictData` or `!`s :stuck_out_tongue:

Comment (jdreaver) [reply library/Stratosphere/ResourceProperties.hs]: Both good points!

Comment (jdreaver) [reply library/Stratosphere/ResourceProperties.hs]: I might just get rid of the `ToJSON` instance for each resource so I don't wrap them up in `Value`.

Comment (jdreaver) [reply library/Stratosphere/ResourceProperties.hs]: Okay I made this `Object`, inlined `ToJSON` for any resources into the `ToResourceProperties` instance, and I also got rid of `TeResourceProperties` instances for resource property types (which aren't resources and shouldn't be able to become resources).

**Comment by utdemir** [comment]
Thank you very much for this PR! I just gave it a go.

* To me, migrating it to not use the `XProperties` type was trivial, as you said just removing those constructors made it compile. I guess it'd only be a problem if someone has a list of `ResourceProperties` or something like that, but I doubt it.

* But the main benefit to me was the file size, size of my executable reduced from 22MB to 11MB! Thanks so much, this solves my issue on #95 .

* Also, just as a test, I ran my application which creates a small stack with 5-ish resources, and it worked fine.

**Comment by jdreaver** [comment]
Thanks for the confirmation @utdemir! I'll run this on our code too. We have a few dozen stacks in production and some tooling to make sure change sets are empty, so once I'm done toying with this PR I'll test it out.

**Comment by jdreaver** [comment]
I ran this on our templates at Freckle and I saw no changes. This looks good to go.

**Pull request merged**